### PR TITLE
interfaces: add xdg-document-portal support to desktop interface

### PR DESCRIPTION
**Pull request by jhenstridge**
This branch builds on top of the work in PR #3963 to expose xdg-document-portal via the desktop interface.

While adding this, I removed some of the default snap-update-ns AppArmor rules so they're only introduced when needed by the desktop interface.

To make sure things are working as intended, I added some new spread tests: one for host font access, and a second for the document portal.

### CODE REVIEW

File: cmd/snap-update-ns/main.go [hunk 1]
```diff
@@ -238,10 +238,7 @@ func applyUserFstab(snapName string) error {
 	}
 
 	// Replace XDG_RUNTIME_DIR in mount profile
-	xdgRuntimeDir, ok := os.LookupEnv("XDG_RUNTIME_DIR")
-	if !ok {
-		return fmt.Errorf("XDG_RUNTIME_DIR is not set")
-	}
+	xdgRuntimeDir := fmt.Sprintf("%s/%d", dirs.XdgRuntimeDirBase, os.Getuid())
```
Comment (zyga) [inline cmd/snap-update-ns/main.go]: Nice :-)

File: cmd/snap-update-ns/main_test.go [hunk 2]
```diff
@@ -288,8 +285,10 @@ func (s *mainSuite) TestApplyUserFstab(c *C) {
 	err = update.ApplyUserFstab("foo")
 	c.Assert(err, IsNil)
 
+	xdgRuntimeDir := fmt.Sprintf("%s/%d", dirs.XdgRuntimeDirBase, os.Getuid())
```
Comment (zyga) [inline cmd/snap-update-ns/main_test.go]: Maybe allow mocking osGetuid for testing? It could be more flexible for testing.

File: interfaces/apparmor/template.go [hunk 3]
```diff
@@ -603,23 +603,8 @@ profile snap-update-ns.###SNAP_NAME### (attach_disconnected) {
   mount options=(ro bind) /var/lib/snapd/hostfs/usr/share/fonts/ -> /snap/###SNAP_NAME###/*/**,
   umount /snap/###SNAP_NAME###/*/**,
 
-  # Allow the desktop interface to bind fonts from the host filesystem
```
Comment (zyga) [inline interfaces/apparmor/template.go]: This looks unrelated.

File: interfaces/builtin/desktop.go [hunk 4]
```diff
@@ -216,8 +220,36 @@ func (iface *desktopInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) b
 	return true
 }
 
+func (iface *desktopInterface) fontconfigDirs() []string {
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: +1 but unrelated to the XDG desktop portal IMO.

File: interfaces/builtin/desktop.go [hunk 5]
```diff
@@ -216,8 +220,36 @@ func (iface *desktopInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) b
...
+	}
+}
+
 func (iface *desktopInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
 	spec.AddSnippet(desktopConnectedPlugAppArmor)
+
+	// Allow mounting fonts
+	for _, dir := range iface.fontconfigDirs() {
+		var buf bytes.Buffer
+		source := "/var/lib/snapd/hostfs" + dir
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: +1 but the constant should gravitate to `dirs.go`

File: interfaces/builtin/desktop.go [hunk 6]
```diff
@@ -216,8 +220,36 @@ func (iface *desktopInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) b
...
+		fmt.Fprintf(&buf, "  mount options=(bind) %s/ -> %s/,\n", source, target)
+		fmt.Fprintf(&buf, "  remount options=(bind, ro) %s/,\n", target)
+		fmt.Fprintf(&buf, "  umount %s/,\n\n", target)
+		spec.AddUpdateNS(buf.String())
+	}
+
+	// Allow mounting document portal
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "  # Mount the document portal\n")
+	fmt.Fprintf(&buf, "  mount options=(bind) /run/user/[0-9]*/doc/by-app/snap.%s/ -> /run/user/[0-9]*/doc/,\n", plug.Snap().Name())
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: ❤️ ❤️ ❤️

File: interfaces/builtin/desktop.go [hunk 7]
```diff
@@ -244,14 +270,12 @@ func (iface *desktopInterface) MountConnectedPlug(spec *mount.Specification, plu
...
-	/*
-		appId := "snap.pkg." + plug.Snap.Name()
-		spec.AddUserMount(mount.Entry{
-			Name: "$XDG_RUNTIME_DIR/doc/by-app/" + appId,
-			Dir: "$XDG_RUNTIME_DIR/doc",
-			Options: []string{"bind", "rw"},
-		})
-	*/
+	appId := "snap." + plug.Snap().Name()
+	spec.AddUserMountEntry(osutil.MountEntry{
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: Can you add a comment explaining what this is doing. Specifically that the document portal doesn't know who's accessing the filesystem and that it needs per-app subdirectories to grok that. I think it should be captured as a comment as otherwise someone might optimise this away eventually.

**Comment by zyga** [comment]
@jhenstridge is this expected? 
```
2018/05/04 11:29:42.370958 main.go:192: cannot change mount namespace of snap "test-snapd-desktop" according to change mount (/run/user/12345/doc/by-app/snap.test-snapd-desktop /run/user/12345/doc none bind,rw 0 0): cannot create path "/run/user/12345/doc": cannot chown path segment "doc" to 0.0 (got up to "/run/user/12345"): operation not permitted
```

File: interfaces/builtin/desktop.go [hunk 8]
```diff
@@ -180,19 +180,20 @@ dbus (send)
...
-#owner /run/user/[0-9]*/doc/** rw,
+owner /run/user/[0-9]*/doc/ r,
+owner /run/user/[0-9]*/doc/** rw,
 
 # Allow access to xdg-desktop-portal and xdg-document-portal
 dbus (receive, send)
     bus=session
     interface=org.freedesktop.portal.*
-    path=/org/freedesktop/portal/{desktop,documents}
+    path=/org/freedesktop/portal/{desktop,documents}{,/**}
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: What are the subsequent components?

File: interfaces/builtin/desktop.go [hunk 9]
```diff
@@ -266,14 +275,12 @@ func (iface *desktopInterface) MountConnectedPlug(spec *mount.Specification, plu
 		})
 	}
 
-	/*
-		appId := "snap.pkg." + plug.Snap.Name()
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: I noticed that the appId has changed, can you explain this? What role does it play?

File: tests/main/interfaces-desktop-document-portal/task.yaml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      echo "Without desktop interface connected"
+      if su -l -c "sh /tmp/check-doc-portal.sh" test 2> check.error; then
+        cat check.error >&2
+        echo "Expected permission error when checking document portal dir"
+        exit 1
+      fi
+      MATCH "Permission denied" < check.error
+    fi
+
+    $LIBEXECDIR/snapd/snap-discard-ns test-snapd-desktop
```
Comment (bboozzoo) [inline tests/main/interfaces-desktop-document-portal/task.yaml]: `"$LIBEXECDIR"/..`

File: tests/main/interfaces-desktop-document-portal/task.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    echo "With desktop connected, we see confined version"
+    snap connect test-snapd-desktop:desktop
+    su -l -c "sh /tmp/check-doc-portal.sh" test | MATCH is-confined
+
+    $LIBEXECDIR/snapd/snap-discard-ns test-snapd-desktop
```
Comment (bboozzoo) [inline tests/main/interfaces-desktop-document-portal/task.yaml]: `"$LIBEXECDIR"/..`

File: tests/main/interfaces-desktop-document-portal/task.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    portal needs to be mounted over $XDG_RUNTIME_DIR/doc inside the
+    sandbox.
+prepare: |
+    snap try $TESTSLIB/snaps/test-snapd-desktop
+    snap disconnect test-snapd-desktop:desktop
+restore: |
+    rm -rf /run/user/12345/*
+    rm -f /tmp/check-doc-portal.sh
+execute: |
+    . $TESTSLIB/dirs.sh
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply tests/main/interfaces-desktop-document-portal/task.yaml]: Thanks for the shellcheck fixes!

**Comment by jhenstridge** [comment]
The current CI failure looks like it a problem with the google:fedora-27-64 spread backend.

    2018-05-11 05:13:36 Error preparing google:fedora-27-64 : 
    -----
    + cat
    + mv gai.conf /etc/gai.conf
    + [[ fedora-27-64 == fedora-* ]]
    + sed -i -s -E -e 's@^#?baseurl=http://download.fedoraproject.org/@baseurl=http://dl.fedoraproject.org/@g' -e 's@^metalink=@#metalink@g' /etc/yum.repos.d/fedora-cisco-openh264.repo /etc/yum.repos.d/fedora-updates-testing.repo /etc/yum.repos.d/fedora-updates.repo /etc/yum.repos.d/fedora.repo
    + dnf --refresh -y makecache
    Fedora 27 - x86_64 - Updates                    666 kB/s |  23 MB     00:35    
    Error: Failed to synchronize cache for repo 'fedora'
    -----

Everything else looks okay.

**Comment by bboozzoo** [comment]
My guess it's Fedora mirrors having bad tiem due to F28 release. There's a PR open, switching F27 to manual, see #5152.

File: interfaces/builtin/desktop.go [hunk 13]
```diff
@@ -250,8 +259,15 @@ func (iface *desktopInterface) AppArmorConnectedPlug(spec *apparmor.Specificatio
 }
 
 func (iface *desktopInterface) MountConnectedPlug(spec *mount.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	appId := "snap." + plug.Snap().Name()
```
Comment (zyga) [inline interfaces/builtin/desktop.go]: I think this should be also limited to classic. On core there's no session yet. Not a blocker though.

File: tests/main/interfaces-desktop-document-portal/task.yaml [hunk 14]
```diff
@@ -0,0 +1,60 @@
+summary: The document portal is mounted by snaps using the desktop interface
+details: |
+    The document portal is a component of xdg-desktop-portal that
+    provides a way to share files with a confined application in a
+    controlled fashion.  To provide proper security, a subtree of the
+    portal needs to be mounted over $XDG_RUNTIME_DIR/doc inside the
+    sandbox.
+systems: [-ubuntu-core-*]
```
Comment (zyga) [inline tests/main/interfaces-desktop-document-portal/task.yaml]: This also proves my point about that this interface feature should be just classic.

File: interfaces/builtin/desktop.go (hunk 13 above)
Comment (zyga) [reply interfaces/builtin/desktop.go]: Oh, that's good then. +1
